src/client/datascience/kernel-launcher/localKernelFinder.ts

```diff
@@ -147,33 +105,38 @@ export class LocalKernelFinder implements ILocalKernelFinder {
         cancelToken?: CancellationToken
     ): Promise<LocalKernelConnectionMetadata[]> {
         try {
-            // Get an id for the workspace folder, if we don't have one, use the fsPath of the resource
-            const workspaceFolderId =
-                this.workspaceService.getWorkspaceFolderIdentifier(
-                    resource,
-                    resource?.fsPath || this.workspaceService.rootPath
-                ) || 'root';
-
-            // If we have not already searched for this resource, then generate the search
-            if (workspaceFolderId && !this.workspaceToMetadata.has(workspaceFolderId)) {
-                this.workspaceToMetadata.set(
-                    workspaceFolderId,
-                    this.findResourceKernelMetadata(resource, cancelToken).then((items) => {
-                        traceInfoIf(
-                            !!process.env.VSC_JUPYTER_LOG_KERNEL_OUTPUT,
-                            `Kernel specs for ${resource?.toString() || 'undefined'} are \n ${JSON.stringify(
-                                items,
-                                undefined,
-                                4
-                            )}`
-                        );
-                        return items;
-                    })
-                );
+            let [nonPythonKernelSpecs, pythonRelatedKernelSpecs] = await Promise.all([
+                this.nonPythonkernelFinder.listKernels(resource, cancelToken),
+                this.extensionChecker.isPythonExtensionInstalled
+                    ? this.pythonKernelFinder.listKernels(resource, cancelToken)
+                    : []
+            ]);
+
+            // If python extension is installed, then ignore Python kernels kernels returned by kernelspec finder.
+            if (!this.extensionChecker.isPythonExtensionInstalled) {
+                nonPythonKernelSpecs = nonPythonKernelSpecs.filter(({ kernelSpec }) => {
+                    if (kernelSpec.language === PYTHON_LANGUAGE) {
```

Comment:
I think this works, but it is possible for non-python kernel specs to be installed in paths relative to Python interpreters. For example IJava does this if you follow their default instructions. But in this case it feels ok to not list them.

Reply:
I probably should rename this file/class to indicate the fact that this only searches the global paths. Probably add comments as well.
The other python finder searches for java kernels found in python interpreters like the one you are referring to.

Open to suggestions for names for the two times

Reply:
Ahh, sorry. I was reading that wrong. You are right that the python extension finder would find them, just the name was throwing me off. Honestly I'm not sure if there is better naming. So:
LocalNonPythonKernelFinder sounds like "finds non-python kernels"
LocalPythonKernelFinder sounds like "finds python kernels" 
Which is sorta right on one level. 

Could be an option:
LocalJupyterKernelFinder => For the global paths might work better? It totally finds python kernels, the tossing out of them happens after they are returned from the finder only if python is install. Then the name of the other class could just stay the same.